### DISCUSSION TITLE
proper Activation Function and Loss Function

### DISCUSSION THREAD

**Discussion by Harris91-lee** [Project-MONAI/MONAI#5009] [Q&A] [answered]
Hello there, 
I'm using SegresNet model in MONAI framwork for multi-label 3D segmentation task and the results were great so far. However, i wonder about the activation function. Since i have multi-label segementation should i change the activation function to softmax? Originally it's set to sigmoid. 
**post_trans originally:**
post_trans = Compose([EnsureType(), Activations(sigmoid=True), AsDiscrete(argmax=True, to_onehot=NUM_OF_LABELS)])
is that correct? 
**Modify:** 
post_trans = Compose([EnsureType(), Activations(softmax=True), AsDiscrete(argmax=True, to_onehot=NUM_OF_LABELS)])
_____________________________
**loss function originally:**
    loss_function = DiceLoss(smooth_nr=0, smooth_dr=1e-5, squared_pred=True, to_onehot_y=False, sigmoid=True)
**Modify:**
　loss_function = DiceLoss(smooth_nr=0, smooth_dr=1e-5, squared_pred=True, to_onehot_y=True, softmax=True)
_____________________________    
i want to say here that even in multi-label segementation using sigmoid didnt effect the results, is that normal? or we have to use softmax for multi-label segmentation task?

**Comment by KumoLiu** [comment, marked as answer]
Hi @Harris91-lee,
Multilabel assumed that classes are not mutually exclusive and each class have its own channel, so use `Sigmoid=True` is correct.

**Reply by Harris91-lee** [reply]
Thank you KumoLiu

**Reply by sh3rlock14** [reply]
Hello, 
I was tackling the same problem, and come out with the same question about using `softmax` (instead of `sigmoid`): in [Brain tumor 3D segmentation with MONAI ](https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/brats_segmentation_3d.ipynb) it's stated that we are dealing with a **multi-labels segmentation** task... but I can't see why/how is that so: given an input volume, shouldn't each voxel/pixel be of exactly one type? *e.g.* background, TC, WT, ET?
Shouldn't we choose exactly one among these possibilities?  

And secondly: I get that the loss functions come with some params to enable an activation function, but I was wondering if I could just combine an activation *e.g.*: `Activations(softmax=True)` and a loss, because I noticed that the `Activations` only compute the activation along `dim 0`, that in my case would be the `batch_dim`, instead of the `labels_dim`. Should I use it only once the batch has been decollated?

**Reply by KumoLiu** [reply]
Hi @sh3rlock14 , 
1. For [brain tumor segmentation ](https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/brats_segmentation_3d.ipynb), you can see that classes are not mutually exclusive, you should use `Sigmoid` instead of `Softmax`. 
2. Before the loss function, you don't need to use `Activations`.

**Reply by sh3rlock14** [reply]
>     1. For [brain tumor segmentation ](https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/brats_segmentation_3d.ipynb), you can see that classes are not mutually exclusive, you should use `Sigmoid` instead of `Softmax`.

I was working on this tutorial but honestly I can't figure it out **how** in this case the classes are **not mutually exclusive**.
I have a clear understanding of 'multi-label' definition in the general case of classification, *e.g.* in Scene Understanding, but for this kind of Segmentation I'am confused:  in each image (let's consider a single volume slice) shouldn't a pixel have only one class? I read online that one should pick `sigmoid` in the case of overlapping regions (as in this case), but none of these comments were proper explanation of the reasoning behind this choice.

I'am still sticking to the fact that we should produce a valid probability distribution over the labels, instead of computing the sigmoid per each one... could you please help me to see why that's not the right way to think about that?

>     2. Before the loss function, you don't need to use `Activations`.

Yeah, sorry: I didn't mean the loss, but actually the metric computation as done in the validation step: in the various tutorials I saw, I noticed that the `Activations` are only used directly on the single input (once the batch has been decollated) rather on the entire batch: do you think letting us pass the dimension onto which compute the *activation* could be more easy?

Thanks in advance for your help!

**Reply by KumoLiu** [reply]
They are not mutually exclusive since each pixel may not point to only one class. You can see `ConvertToMultiChannelBasedOnBratsClassesd` for how they get the label.

**Reply by sh3rlock14** [reply]
Thanks for your feedback:
I actually wrote my version of `ConvertToMultiChannelBasedOnBratsClassesd` since I noticed some discrepancies in the one implemented in the [tutorial](https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/brats_segmentation_3d.ipynb), and the one in the [docs](https://docs.monai.io/en/stable/_modules/monai/transforms/utility/array.html#ConvertToMultiChannelBasedOnBratsClasses), in particular:

Tutorial:
- label 1 is the peritumoral edema
- label 2 is the GD-enhancing tumor
- label 3 is the necrotic and non-enhancing tumor core

Library:
- label 1 is the necrotic and non-enhancing tumor core
- label 2 is the peritumoral edema
- label 4 is the GD-enhancing tumor

while after downloading the `Task01_BrainTumour` dataset, I found out that the labels were:

- label 0 is the background
- label 1 is the peritumoral edema 
- label 2 is the non-enhancing tumour
- label 3 is the enhancing tumour

So, my version is:

```
class ToMultiChannelBratsClassesd(MapTransform):
    """
    Convert labels to multi channels based on brats classes
    label 1 is the peritumoral edema
    label 2 is the GD-enhancing tumor
    label 3 is the necrotic and non-enhancing tumor core

    The classes are: BG (background), TC (Tumore core), WT (Whole tumor), ET (Enhancing tumor)
    """
    def __call__(self, data):
        d = dict(data)
        for key in self.keys:
            # merge label 2 and 3 to consrtuct TC
            # merge label 1, 2 and 3 to construct WT
            # label 2 is ET
            result = [ (d[key]==0), (d[key]==2) | (d[key]==3), (d[key]==1) | (d[key]==2) | (d[key]==3), d[key]==2 ]
            d[key] = stack(result, axis=0).float()
        return d
```

and I do not perform any kind of target inversion while computing the loss: given an `input` of size `BCHWD`, the model outputs a tensor `BNHWD`, where `N=4` (one channel for each class).

Then I compute:
```
criterion = DiceLoss(include_background=True,
                         to_onehot_y=False,
                         sigmoid=False,
                         softmax=True,
                         ...
                         )

dice_loss = criterion(output, target)
``` 

So, if I understand correctly your prompt, one should use `sigmoid=True` since the **classes** (TC, WT, ET) defined here are actually a merging of multiple labels?

For my own purposes I think I'll try both configurations, but do you think that going with the `softmax` is conceptually wrong?

**Reply by KumoLiu** [reply]
Yes, if `softmax` is used, then the sum of the probabilities of each label will become 1, which is not correct.
